src/core/trulens/experimental/otel_tracing/core/otel.py

```diff
@@ -0,0 +1,778 @@
+from opentelemetry import trace as trace_api
+from opentelemetry.sdk import resources as resources_sdk
+from opentelemetry.sdk import trace as trace_sdk
+from opentelemetry.util import types as types_api
+import pydantic
+from trulens.core.utils import python as python_utils
+from trulens.core.utils import serial as serial_utils
+from trulens.core.utils.python import TypeAlias
+from trulens.core.utils.python import TypeAliasType
+from trulens.core.utils.serial import Lens
+
+logger = logging.getLogger(__name__)
+
+# Type alises
+
+A = TypeVar("A")
+B = TypeVar("B")
+
+TSpanID: TypeAlias = int
+"""Type of span identifiers.
+64 bit int as per OpenTelemetry.
+"""
+NUM_SPANID_BITS: int = 64
+"""Number of bits in a span identifier."""
+
+TTraceID: TypeAlias = int
+"""Type of trace identifiers.
+128 bit int as per OpenTelemetry.
+"""
+NUM_TRACEID_BITS: int = 128
+"""Number of bits in a trace identifier."""
+
+TTimestamp: TypeAlias = int
+"""Type of timestamps in spans.
+
+64 bit int representing nanoseconds since epoch as per OpenTelemetry.
+"""
+NUM_TIMESTAMP_BITS = 64
+
+TLensedBaseType: TypeAlias = Union[str, int, float, bool]
+"""Type of base types in span attributes.
+
+!!! Warning
+    OpenTelemetry does not allow None as an attribute value. Handling None is to
+    be decided.
+"""
+
+TLensedAttributeValue = TypeAliasType(
+    "TLensedAttributeValue",
+    Union[
+        str,
+        int,
+        float,
+        bool,
+        python_utils.NoneType,  # NOTE(piotrm): None is not technically allowed as an attribute value.
+        Sequence["TLensedAttributeValue"],  # type: ignore
+        "TLensedAttributes",
+    ],
+)
+"""Type of values in span attributes."""
+
+# NOTE(piotrm): pydantic will fail if you specify a recursive type alias without
+# the TypeAliasType schema as above.
+
+TLensedAttributes: TypeAlias = Dict[str, TLensedAttributeValue]
+"""Attribute dictionaries.
+
+Note that this deviates from what OTEL allows as attribute values. Because OTEL
+does not allow general recursive values to be stored as attributes, we employ a
+system of flattening values before exporting to OTEL. In this process we encode
+a single generic value as multiple attributes where the attribute name include
+paths/lenses to the parts of the generic value they are representing. For
+example, an attribute/value like `{"a": {"b": 1, "c": 2}}` would be encoded as
+`{"a.b": 1, "a.c": 2}`. This process is implemented in the
+`flatten_lensed_attributes` method.
+"""
+
+
+def flatten_value(
+    v: TLensedAttributeValue, lens: Optional[Lens] = None
+) -> Iterable[Tuple[Lens, types_api.AttributeValue]]:
+    """Flatten recursive value into OTEL-compatible attribute values.
+
+    See `TLensedAttributes` for more details.
+    """
+
+    if lens is None:
+        lens = Lens()
+
+    # OpenTelemetry does not allow None as an attribute value. Unsure what
+    # is best to do here.
+    # if v is None:
+    #    yield (path, "None")
```

Comment:
Please file a TODO jiras for these under otel epic

Reply:
Did this and some others.